modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/servlet/taglib/ScopedCSSVariablesTopHeadDynamicInclude.java

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.css.variables.web.internal.servlet.taglib;
+
+import com.liferay.frontend.css.variables.ScopedCSSVariables;
+import com.liferay.frontend.css.variables.ScopedCSSVariablesProvider;
+import com.liferay.portal.kernel.servlet.taglib.BaseDynamicInclude;
+import com.liferay.portal.kernel.servlet.taglib.DynamicInclude;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.osgi.service.component.annotations.ReferencePolicyOption;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+@Component(immediate = true, service = DynamicInclude.class)
+public class ScopedCSSVariablesTopHeadDynamicInclude
+	extends BaseDynamicInclude {
+
+	@Override
+	public void include(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse, String dynamicIncludeKey)
+		throws IOException {
+
+		PrintWriter printWriter = httpServletResponse.getWriter();
+
+		printWriter.print("<style data-senna-track=\"temporary\" ");
+		printWriter.print("type=\"text/css\">\n");
+
+		for (ScopedCSSVariablesProvider scopedCSSVariablesProvider :
+				_scopedCssVariablesProviders) {
+
+			_writeCSSVariables(
+				printWriter,
+				scopedCSSVariablesProvider.getScopedCSSVariablesCollection(
+					httpServletRequest));
+		}
+
+		printWriter.print("</style>\n");
+	}
+
+	@Override
+	public void register(DynamicIncludeRegistry dynamicIncludeRegistry) {
+		dynamicIncludeRegistry.register(
+			"/html/common/themes/top_head.jsp#post");
```

Comment:
I guess it doesn't really matter, but would it make more sense to use `top_head.jsp#pre` for the variable declarations?

Reply:
Yes it matters and no it wouldn't make more sense :sweat_smile: .

We need to place the CSS vars after the theme's CSS so that they take precedence. It is described in [this commit](https://github.com/liferay-frontend/liferay-portal/pull/21/commits/66840fe6121b60f415e8398c44f7f3e54d095571), though it is not very clear :roll_eyes: .

Reply:
Hm. Got it. Thx